src/components/Dialogs/Conditions/OptionSetCondition.js

```diff
@@ -0,0 +1,205 @@
+import PropTypes from 'prop-types'
+import React, { useState } from 'react'
+import { connect } from 'react-redux'
+import { apiFetchOptions } from '../../../api/options.js'
+import { OPERATOR_IN } from '../../../modules/conditions.js'
+import { useDebounce, useDidUpdateEffect } from '../../../modules/utils.js'
+import { sGetMetadata } from '../../../reducers/metadata.js'
+import { sGetSettingsDisplayNameProperty } from '../../../reducers/settings.js'
+import classes from './styles/Condition.module.css'
+
+const LeftHeader = ({ searchTerm, setSearchTerm, dataTest }) => (
+    <div className={classes.transferLeftHeader}>
+        <p className={classes.transferLeftTitle}>
+            {i18n.t('Available options')}
+        </p>
+        <InputField
+            value={searchTerm}
+            onChange={({ value }) => setSearchTerm(value)}
+            placeholder={i18n.t('Filter options')}
+            dataTest={`${dataTest}-filter-input-field`}
+            dense
+            initialFocus
+            type={'search'}
+        />
+    </div>
+)
+
+LeftHeader.propTypes = {
+    dataTest: PropTypes.string,
+    searchTerm: PropTypes.string,
+    setSearchTerm: PropTypes.func,
+}
+
+const EmptySelection = () => (
+    <p className={classes.transferEmptyList}>{i18n.t('No items selected')}</p>
+)
+const RightHeader = () => (
+    <p className={classes.transferRightHeader}>{i18n.t('Selected options')}</p>
+)
+
+const SourceEmptyPlaceholder = ({ loading, searchTerm, options, dataTest }) =>
+    !loading &&
+    !options.length && (
+        <p className={classes.transferEmptyList} data-test={dataTest}>
+            {searchTerm
+                ? i18n.t('Nothing found for "{{- searchTerm}}"', {
+                      searchTerm: searchTerm,
+                  })
+                : i18n.t('No options')}
+        </p>
+    )
+
+SourceEmptyPlaceholder.propTypes = {
+    dataTest: PropTypes.string,
+    loading: PropTypes.bool,
+    options: PropTypes.array,
+    searchTerm: PropTypes.string,
+}
+
+const OptionSetCondition = ({
+    condition,
+    optionSetId,
+    onChange,
+    displayNameProp,
+    metadata,
+}) => {
+    const parts = condition.split(':')
+    const values = parts[1]?.length ? parts[1].split(';') : []
+    const selectedOptions = values.map((code) => ({
+        code,
+        name: metadata[code]?.name, // FIXME: Doesn't work as metadata stores the options with an id that's separate from the code
+    }))
+    const dataTest = 'option-set'
+
+    const setValues = (selected) => {
+        onChange(`${OPERATOR_IN}:${selected.join(';') || ''}`)
+    }
+
+    const [state, setState] = useState({
+        searchTerm: '',
+        options: [],
+        loading: true,
+        nextPage: 1,
+    })
+    const dataEngine = useDataEngine()
+    const setSearchTerm = (searchTerm) =>
+        setState((state) => ({ ...state, searchTerm }))
+    const debouncedSearchTerm = useDebounce(state.searchTerm, 200)
+    const fetchItems = async (page) => {
```

Comment:
Perhaps out of scope but there is a way of doing this with `useDataQuery`.
Example here: https://github.com/dhis2/ui/blob/master/components/sharing-dialog/src/autocomplete/sharing-autocomplete.js